addons/middle-click-popup/userscript.js

```diff
@@ -0,0 +1,469 @@
+      this.floatInput.className = "sa-float-bar-input";
+      this.floatInput.className = addon.tab.scratchClass("input_input-form", {
+        others: "sa-float-bar-input",
+      });
+
+      this.dropdown = this.dropdownOut.appendChild(document.createElement("ul"));
+      this.dropdown.className = "sa-float-bar-dropdown";
+
+      this.floatInput.addEventListener("keyup", () => this.inputChange());
+      this.floatInput.addEventListener("focus", () => this.inputChange());
+      this.floatInput.addEventListener("keydown", (...e) => this.inputKeyDown(...e));
+      this.floatInput.addEventListener("focusout", () => this.hide());
+
+      this.dropdownOut.addEventListener("mousedown", (...e) => this.onClick(...e));
+
+      document.addEventListener("keydown", (e) => {
+        let ctrlKey = e.ctrlKey || e.metaKey;
+
+        if (e.key === " " && ctrlKey) {
+          // Ctrl + Space (Inject Code)
+          this.show(e);
+          e.cancelBubble = true;
+          e.preventDefault();
+          return true;
+        }
+      });
+    }
+
+    show(e) {
+      if (this.selectedTab !== 0) {
+        return;
+      }
+
+      e.cancelBubble = true;
+      e.preventDefault();
+
+      this.buildFilterList();
+
+      this.floatBar.style.left = (e.clientX ?? mouse.x) + 16 + "px";
+      this.floatBar.style.top = (e.clientY ?? mouse.y) - 8 + "px";
+      this.floatBar.style.display = "";
+      this.floatInput.value = "";
+      this.floatInput.focus();
+    }
+
+    onClick(e) {
+      e.cancelBubble = true;
+      if (!e.target.closest("input")) {
+        e.preventDefault();
+      }
+
+      let sel = e && e.target;
+      if (sel.tagName === "B") {
+        sel = sel.parentNode;
+      }
+
+      if (e instanceof MouseEvent && sel.tagName !== "LI") {
+        // Mouse clicks need to be on a block...
+        return;
+      }
+
+      if (!sel || !sel.data) {
+        sel = this.dropdown.querySelector(".sel");
+      }
+
+      if (!sel) {
+        return;
+      }
+
+      let option = sel.data.option;
+      // block:option.block, dom:option.dom, option:option.option
+      if (option.option) {
+        // We need to tweak the dropdown in this xml...
+        let field = option.dom.querySelector("field[name=" + option.pickField + "]");
+        if (field.getAttribute("id")) {
+          field.innerText = option.option[0];
+          field.setAttribute("id", option.option[1] + "-" + option.option[0]);
+        } else {
+          field.innerText = option.option[1]; // griffpatch - oops! option.option[1] not 0?
+        }
+
+        // Handle "stop other scripts in sprite"
+        if (option.option[1] === "other scripts in sprite") {
+          option.dom.querySelector("mutation").setAttribute("hasnext", "true");
+        }
+      }
+
+      this.createDraggingBlock(option.dom, e, sel);
+
+      if (e.shiftKey) {
+        this.floatBar.style.display = "";
+        this.floatInput.focus();
+      }
```

Comment:
Nitpick, but it would probably be better to move this all into `createDraggingBlock` than instead of calling `onClick` from `inputKeyDown` you could just call `createDraggingBlock` instead. It would make it a bit clearer what's happening